---
title: "Analyze a single-dataset with Scriabin"
author: "Aaron J. Wilk"
date: "2021-12-15"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{single-dataset}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

Here we will illustrate the first Scriabin workflow: generating and analyzing a full dataset cell-cell interaction matrix. We choose the panc8 dataset of pancreas islets that can be easily installed via SeuratData. 

Because one dimension of the cell-cell interaction matrix is the (number of cells in the dataset)^2, it is impractical to generate and analyze this matrix for medium-to-large datasets. In these situations, we recommend trying the other two workflows of Scriabin. Namely, identification of cells whose magnitude of communication 

Load libraries

```{r}
library(Seurat)
library(SeuratData)
library(scriabin)
library(tidyverse)
```

To install the panc8 dataset:
```{r}
if (!requireNamespace("panc8.SeuratData", quietly = TRUE))
  install.packages("https://seurat.nygenome.org/src/contrib/panc8.SeuratData_3.0.2.tar.gz", repos = NULL, type = "source") 
library(panc8.SeuratData)
panc8 <- LoadData("panc8")
```

Analyze the fluidigmc1 dataset for the single dataset tutorial

Analyze the indrop dataset for the interaction programs tutorial
```{r}
panc_fl <- subset(panc8, cells = colnames(panc8)[panc8$tech=="fluidigmc1"])
panc_fl <- SCTransform(panc_fl, verbose = F) %>%
  RunPCA(verbose = F) %>%
  RunUMAP(dims = 1:30, verbose = F)
DimPlot(panc_fl, group.by = "celltype", label = T, repel = T) + NoLegend()
```

Is there communication heterogeneity between alpha and beta cells?

First we generate a cell-cell interaction matrix object on the alpha and beta cells in the dataset. Then we scale, find variable features, run PCA and UMAP, find neighbor graphs and clusters, treating this matrix just like we would a gene expression matrix.

```{r}
panc_fl_ccim <- GenerateCCIM(subset(panc_fl, cells = colnames(panc_fl)[panc_fl$celltype %in% c("alpha","beta")]))
panc_fl_ccim <- ScaleData(panc_fl_ccim) %>% 
  FindVariableFeatures() %>%
  RunPCA() %>% 
  RunUMAP(dims = 1:10) %>%
  FindNeighbors(dims = 1:10) %>%
  FindClusters(resolution = 0.2)

p1 <- DimPlot(panc_fl_ccim, group.by = "sender_celltype")
p2 <- DimPlot(panc_fl_ccim, group.by = "receiver_celltype")
p1+p2

DimPlot(panc_fl_ccim, label = T, repel = T) + NoLegend()
```

First, there's some interesting heterogeneity here: alpha-beta or beta-alpha sender-receiver pairs do not all appear to interact in the same way. Further, we visually see strong separation between cell-cell pairs with different sender-receiver cell type combinations. 

Some clusters of cell-cell pairs with the same sender annotation contain receivers of different annotations (eg. cluster 0), perhaps indicating these sender cells use similar patterns of ligand expression to communicate with both cell types. The converse is true of other clusters (eg. cluster 6). 

Next, we can identify what ligand-receptor pairs are differentially expressed between these clusters of CCC by calling FindMarkers. Clusters 4 and 5 look particularly curious, where in both clusters alpha cells are the receivers from either alpha (4) or beta (5) cells.

```{r}
cluster4_5.edges <- FindMarkers(panc_fl_ccim, ident.1 = "4", ident.2 = "5")

cluster4_5.edges %>% top_n(20,wt = abs(avg_log2FC))
```

Scriabin also provides a utility function CCIMFeaturePlot to visualize the expression of ligands or receptors in gene expression space superimposed on cell-cell pairs.

```{r}
CCIMFeaturePlot(panc_fl_ccim, seu = panc_fl, 
            features = c("EFNA1","NPNT"), type_plot = "sender")
```



